### PR TITLE
feature/filter-engine-as-a-service

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Description
-----------
Have the filter engine as a service

Impacts
-----------
The Admin

NEWS file entries
-------

New Features
++++++++++++
* A new service pffilter

Delete branch after merge
-------------------------
NO

### CODE REVIEW

File: conf/documentation.conf [hunk 1]
```diff
@@ -623,6 +630,13 @@ Order in which the service should be started.
 Lower order services start first and stop last.
 EOT
 
+[services.pffilter_order]
```
Comment (extrafu) [inline conf/documentation.conf]: Do we need that? I don't think so, especially with systemd.

File: conf/log.conf.d/pffilter.conf.example [hunk 2]
```diff
@@ -0,0 +1,18 @@
+### pffilter logger ###
+log4perl.rootLogger = INFO, PFFILITER
```
Comment (extrafu) [inline conf/log.conf.d/pffilter.conf.example]: filiter?

File: conf/log.conf.d/pffilter.conf.example [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+### Categories ###
+# Below, you can specify different categories (based on package names) for different logging levels
+#log4perl.category.pf.SNMP = WARN
+
+### Logging for pffilter
+log4perl.appender.PFFILITER                              = pf::log::FileAppender
```
Comment (extrafu) [inline conf/log.conf.d/pffilter.conf.example]: Filiter? everywhere.

File: conf/documentation.conf (hunk 1 above)
Comment (louismunro) [reply conf/documentation.conf]: Ordering is obsolete.
It will be removed any minute now.

File: lib/pf/Switch.pm [hunk 4]
```diff
@@ -3269,6 +3269,19 @@ sub externalPortalEnforcement {
 }
 
 
+=item TO_JSON
+
+TO_JSON
+
+=cut
+
+sub TO_JSON {
```
Comment (extrafu) [inline lib/pf/Switch.pm:3278]: Useful? It doesn't seem to be called anywhere and if it is, why in all caps?

**Comment by jrouzierinverse** [comment]
It is special named method for the JSON encoder to properly handle 
objects when encoding them to json.

This was there so the perl pffilter client would work when we trying to 
send a switch info in it's payload

James Rouzier
jrouzier@inverse.ca :: +1.514.447.4918 (x115)  ::  http://www.inverse.ca
Inverse inc. :: Leaders behind SOGo (http://www.sogo.nu) and PacketFence (http://www.packetfence.org)

File: go/filter_client/filter_client.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
+package filter_client
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"net"
+)
+
+const default_socket_path string = "/usr/local/pf/var/run/pffilter.sock"
```
Comment (julsemaan) [inline go/filter_client/filter_client.go]: We try to adopt the standard for casing variables which is camel case

File: lib/pf/pffilter/client.pm [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+sub send_request {
+    my ($self, $method, $params) = @_;
+    my $socket = $self->get_socket;
+    my %request  = (
+        method => $method,
+        params => $params
+    );
+    my $bytes = encode_json(\%request);
+#    my $bytes = $JSON->encode(\%request);
```
Comment (julsemaan) [inline lib/pf/pffilter/client.pm]: not useful it seems

File: lib/pf/pffilter/client.pm [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    print $socket $bytes . "\n";
+    $bytes = <$socket>;
+    if (!defined $bytes) {
+        die "Error ";
+    }
+    if ($bytes) {
+#        my $response = $JSON->decode($bytes);
```
Comment (julsemaan) [inline lib/pf/pffilter/client.pm]: neither this

File: sbin/pffilter [hunk 8]
```diff
@@ -0,0 +1,533 @@
+#!/usr/bin/perl
+
+=head1 NAME
+
+pffilter - ARP listener and maintenance threads
```
Comment (julsemaan) [inline sbin/pffilter]: ARP listener ?

File: sbin/pffilter [hunk 9]
```diff
@@ -0,0 +1,533 @@
...
+my $RADIUS_FILTER = pf::access_filter::radius->new;
+my $DHCP_FILTER = pf::access_filter::dhcp->new;
+my $DNS_FILTER = pf::access_filter::dns->new;
+my $VLAN_FILTER = pf::access_filter::vlan->new;
+
+pf::SwitchFactory->preloadConfiguredModules();
+
+# initialization
+# --------------
+# assign process name (see #1464)
```
Comment (julsemaan) [inline sbin/pffilter]: you're not assigning a process name

File: go/filter_client/filter_client.go (hunk 5 above)
Comment (jrouzierinverse) [reply go/filter_client/filter_client.go]: Is our golang coding style documented somewhere?

Comment (julsemaan) [reply go/filter_client/filter_client.go]: No, I believe we currently just refer to the known Golang standards for now

Unless @louismunro had fun doing it

**Comment by jrouzierinverse** [comment]
Done

**Pull request merged**